### PR TITLE
Clarify service worker timeout

### PR DESCRIPTION
**Pull request by jpmedley**
Fixes #SOME_ISSUE_NUMBER

Changes proposed in this pull request:

-
-
-

### CODE REVIEW

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 1]
```diff
@@ -37,7 +38,7 @@ Unfortunately, this behavior did not apply to extension events. Extension events
 
 As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events.
 
-While this change removes the five minute limit on extension service worker lifetimes, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
+Because the service worker will still terminate after five minutes if there are no events, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
```
Comment (rdcronin) [inline site/en/blog/longer-esw-lifetimes/index.md:41]: This still isn't quite right.
- The service worker will terminate after 30 seconds if there are no events
- The service worker will terminate after 5 minutes if a single request (event or API call) takes longer than 5 minutes to process

**Review by rdcronin — COMMENTED** [review]
Thanks, Joe!  I think we still need to tweak this a bit.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 1 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md]: PTAL

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 2]
```diff
@@ -35,9 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
...
 
-As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events.
+As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events. Be aware that:
 
-While this change removes the five minute limit on extension service worker lifetimes, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
+* The service worker will terminate after 30 seconds if there are no events.
+* The service worker will terminate after 5 minutes if a single request (event or API call) takes longer than 5 minutes to process.
```
Comment (rdcronin) [inline site/en/blog/longer-esw-lifetimes/index.md:42]:
```suggestion
* The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process (with the exception of a native messaging connection).
```

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 3]
```diff
@@ -35,9 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
 
 ## What's changed
 
-As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events.
+As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events. Be aware that:
```
Comment (rdcronin) [inline site/en/blog/longer-esw-lifetimes/index.md:39]:
```suggestion
As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. Be aware that:
```

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 4]
```diff
@@ -35,9 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
```
Comment (rdcronin) [inline site/en/blog/longer-esw-lifetimes/index.md:36]: Github won't let me comment on line 33, but:

> New service worker events received before the idle timer or long running task timer expire would reset the timers and extend the service worker's lifetime.

This should read "... would reset the idle timer and..." vs "reset the timers and".  The long-running task timer is not reset.

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 5]
```diff
@@ -35,9 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
...
+* The service worker will terminate after 30 seconds if there are no events.
+* The service worker will terminate after 5 minutes if a single request (event or API call) takes longer than 5 minutes to process.
+
+You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
 
 In a future release, we may change how specific events are processed to prevent keeping service workers alive indefinitely, but without shutting down long-running work. In the meantime, make sure your extensions yield when possible and prepare for termination by persisting extension states.
```
Comment (rdcronin) [inline site/en/blog/longer-esw-lifetimes/index.md:46]: This line doesn't make sense to me.

> to prevent keeping service workers alive indefinitely

We don't mind this as long as it's legitimately useful work.

> but without shutting down long-running work

This seems directly contradictory to the long-running task timer, which explicitly _will_ shut down long-running work.

If you want to keep this sentence, maybe:

"Future releases of Chromium may prevent artificially extending service worker lifetimes."

That said, I we could even just remove this whole sentence, and just append the second to the previous paragraph:

"... you should test your extensions to ensure that they're not doing this unintentionally and make sure your extensions yield whenever possible..."

**Review by rdcronin — CHANGES_REQUESTED** [review]

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 4 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:36]: Done.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 5 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:46]: I took the second approach. The existing language was Simeon's. Not only do I prefer to stay away from remarks about the future, such promises are also contrary to Google's style guide.

**Comment by jpmedley** [comment]
@rdcronin, back to you. I hope I got it this time.

**Comment by jpmedley** [comment]
@rdcronin Can you please give this a final pass. Thanks.

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 6]
```diff
@@ -35,11 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
...
+As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. Be aware that:
 
-While this change removes the five minute limit on extension service worker lifetimes, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
+* The service worker will terminate after 30 seconds if there are no events.
+* The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process (with the exception of a native messaging connection).
```
Comment (sebastianbenz) [inline site/en/blog/longer-esw-lifetimes/index.md:42]: I assume this is only the case for pending async calls, correct? Or will synchronous operations be terminated as well?

//cc @rdcronin

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 7]
```diff
@@ -35,11 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
...
 
-In a future release, we may change how specific events are processed to prevent keeping service workers alive indefinitely, but without shutting down long-running work. In the meantime, make sure your extensions yield when possible and prepare for termination by persisting extension states. 
+You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally. Make sure your extensions yield when possible and prepare for termination by persisting extension states.
```
Comment (sebastianbenz) [inline site/en/blog/longer-esw-lifetimes/index.md:44]: This does not quite fit the rest of the flow: "You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination."

Reading the post I see nothing that would indicate that sw's are no longer event-based (as we don't explicitly mention hacking the lifecycle).

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 8]
```diff
@@ -35,11 +36,12 @@ Unfortunately, this behavior did not apply to extension events. Extension events
 
 ## What's changed
 
-As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events.
+As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. Be aware that:
```
Comment (sebastianbenz) [inline site/en/blog/longer-esw-lifetimes/index.md:39]: I think it'd be great to elaborate a bit more on this:

- why are we're making this change?
- what was the previous behavior? (this helps devs gauge how this might impact the behavior of their extension)

Another thing: can we either remove `typically` or make it more precise? left on its own it is a bit confusing (at least I stumbled over it).

**Review by sebastianbenz — COMMENTED** [review]
Looking good! Glad to see this go out. Left a few comments.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 8 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:39]: >why are we're making this change?

Because premature termination was causing errors in people's code. At the time we figured that if somebody wasn't  having that problem they wouldn't be reading this article. Frankly, I'd like to err on the side of getting this out.

>what was the previous behavior?

It's described in the background.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 6 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:42]: I don't know. Maybe Devlin does.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 7 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:44]: I remember no. Simeon strongly insisted on this line. We can get rid of it. @oliverdunk I'd be curious to know, if you have any opinions on this line.

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 9]
```diff
@@ -16,6 +17,10 @@ alt: >
   A picture of an hourglass.
 ---
 
+{% Aside %}
+This post has been updated since we originally published it in January. TBD -summary of changes.
```
Comment (jpmedley) [inline site/en/blog/longer-esw-lifetimes/index.md:21]: @sebastianbenz I'll fill this out when the content is settled.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 7 above)
Comment (oliverdunk) [reply site/en/blog/longer-esw-lifetimes/index.md:48]: I think the two important things to communicate are:

- We may change this behaviour to be stricter in the future (though hopefully in a way that does not impact genuine extensions).
- Your extension may terminate unexpectedly even with these changes (for example, if the browser crashes).

How about something like:

> In a future release, we may continue to iterate on this behaviour to ensure extension service workers are terminated when possible, without shutting down long-running work. As part of building a resource conscious extension, extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This makes sure they are robust against unpredictable events like the browser being forcefully closed by the user.

I think that's a bit wordy but better in some sense?

Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md]: I just submitted a change with a version of Oliver's suggestion.

File: site/en/blog/longer-esw-lifetimes/index.md (hunk 6 above)
Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:46]: New wording from Oliver.

Comment (jpmedley) [reply site/en/blog/longer-esw-lifetimes/index.md:46]: Oliver didn't like the new wording and he makes some compelling arguments:
* >It won't terminate after 30 seconds if you make certain API calls for example. So it would need to be "The service worker will terminate after 30 seconds if there are no events or API calls to extension APIs and there isn't a native messaging connection, regardless of whether you're running synchronous or asynchronous code." I realise we touch on that elsewhere in the article but I think we should make sure if people quote a specific bullet point it isn't wrong.
* >I feel like synchronous and asynchronous was sort of implied with how we had it before, and calling it out actually adds confusion because then it makes me stop to think about what those are.

Comment (oliverdunk) [reply site/en/blog/longer-esw-lifetimes/index.md:46]: I'm not sure what the latest wording is but I think something like this would be fine:

> * The service worker will terminate after 30 seconds of inactivity (receiving an event or calling an extension API resets this timer).
> * The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process
>
> Some APIs like native messaging provide a strong keep-alive which cancel both of these timers.

And maybe change "be aware that" to "specifically" or something? Since it's not a bad thing.

Comment (sebastianbenz) [reply site/en/blog/longer-esw-lifetimes/index.md:46]: +1 to Oliver's suggestions

**Comment by jpmedley** [comment]
@sebastianbenz PTAL

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 10]
```diff
@@ -16,6 +17,10 @@ alt: >
   A picture of an hourglass.
 ---
 
+{% Aside %}
+This post has been updated since we originally published it in January. We've added text indicating that calls to extension APIs reset the service worker's timeout clock, and clarified exactly when service worker timeouts will ocurr.
```
Comment (oliverdunk) [inline site/en/blog/longer-esw-lifetimes/index.md:21]: Hopefully this is the US spelling too:

```suggestion
This post has been updated since we originally published it in January. We've added text indicating that calls to extension APIs reset the service worker's timeout clock, and clarified exactly when service worker timeouts will occur.
```

File: site/en/blog/longer-esw-lifetimes/index.md [hunk 11]
```diff
@@ -35,11 +40,14 @@ Unfortunately, this behavior did not apply to extension events. Extension events
...
 
-While this change removes the five minute limit on extension service worker lifetimes, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.
+Some APIs like native messaging provide a strong keep-alive which cancel both of these timers.
 
-In a future release, we may change how specific events are processed to prevent keeping service workers alive indefinitely, but without shutting down long-running work. In the meantime, make sure your extensions yield when possible and prepare for termination by persisting extension states. 
+We are till working to ensure that extension service workers are terminated when possible, without shutting down long-running work. A resource-concious extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This ensures against unpredictable events like the browser being forcefully closed by the user.
```
Comment (oliverdunk) [inline site/en/blog/longer-esw-lifetimes/index.md:50]:
```suggestion
We are still working to ensure that extension service workers are terminated when possible, without shutting down long-running work. Resource-concious extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This guards against unpredictable events like the browser being forcefully closed by the user.
```

**Review by oliverdunk — APPROVED** [review]
Left a couple of nits but otherwise this looks good to me (would still love for Sebastian to approve as well) 😄

**Review by sebastianbenz — APPROVED** [review]
LGTM once Oliver's feedback has been addressed.

**Comment by jpmedley** [comment]
I'm getting build issues I can't resolve. Opening a new PR.
